### PR TITLE
Add Prometheus monitoring.

### PR DESCRIPTION
**Pull request by jsha**
Note: Right now Prometheus is mainly monitoring itself. We need to
figure out how to push metrics from our cron jobs, since Prometheus
can't scrape them (by virtue of being cron jobs).

### CODE REVIEW

File: terraform/modules/monitoring/monitoring.tf
Comment (tgeoghegan) [inline terraform/modules/monitoring/monitoring.tf:70]: Is this role not namespaced? There's a distinction between a `kubernetes_cluster_role` and a [`kubernetes_role`](https://registry.terraform.io/providers/hashicorp/kubernetes/latest/docs/resources/role).

File: terraform/modules/monitoring/monitoring.tf
Comment (tgeoghegan) [inline terraform/modules/monitoring/monitoring.tf:129]: nit: `prometheus_store`?

File: terraform/modules/monitoring/monitoring.tf
Comment (tgeoghegan) [inline terraform/modules/monitoring/monitoring.tf:133]: If the backing store for Prometheus data is in `$REGION-a`, what does that mean about the availability of Prometheus? Can it only run in zone a?

File: terraform/modules/monitoring/monitoring.tf
Comment (tgeoghegan) [inline terraform/modules/monitoring/monitoring.tf:63]: nit: it's a shame having some of this farmed out to a separate YAML file, especially since some of the values in there need to agree with values here in Terraform (e.g., the port numbers). I think you could use [yamlencode](https://www.terraform.io/docs/configuration/functions/yamlencode.html) to instead define all those parameters inline in Terraform.

**Review by tgeoghegan — DISMISSED** [review]

File: terraform/modules/monitoring/monitoring.tf
Comment (jsha) [reply terraform/modules/monitoring/monitoring.tf:70]: Correct, this is meant to be a cluster role (not a role) because it has to be non-namespaced in order to reach into all the namespaces.

File: terraform/modules/monitoring/monitoring.tf
Comment (jsha) [reply terraform/modules/monitoring/monitoring.tf:63]: FWIW, I am more used to maintaining Prometheus config as its own YAML, and prefer it that way. For instance if we need to copy a config block from some example, we don't need to transmute it into HCL first.

The one parameter that needs agreement is the port, and 9090 is a standard port for Prometheus, so I'm not worried about writing it twice.

**Comment by jsha** [comment]
Updates with some fixes to make this actually use a persistent store. I'll wait for a re-review before merging.

File: terraform/modules/monitoring/monitoring.tf
Comment (yuriks) [reply terraform/modules/monitoring/monitoring.tf:179]: GCP supports regional disks (they're just potentially slightly slower, but not by that much), so I think we should just use that for better availability. Unless those don't work here for some reason?

Comment (yuriks) [reply terraform/modules/monitoring/monitoring.tf:179]: Ah, seems like they're a just a different type of resource in the API (and therefore TF), so you'd miss that if just looking at `google_compute_disk`: https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/compute_region_disk

File: terraform/modules/monitoring/monitoring.tf
Comment (yuriks) [inline terraform/modules/monitoring/monitoring.tf:175]: You should probably add `provider = google-beta` or this will use incorrect defaults (because we don't provide configuration for the default non-beta `google` one) and that tends to cause confusing error messages. (I plan on PRing a fix for this in general soon though.)

File: terraform/modules/monitoring/monitoring.tf
Comment (tgeoghegan) [inline terraform/modules/monitoring/monitoring.tf:30]: Could you add a comment explaining what is exposed on these ports? Is this the Prometheus API?

File: terraform/modules/monitoring/monitoring.tf
Comment (tgeoghegan) [inline terraform/modules/monitoring/monitoring.tf:49]: Is pulling `latest` safe? Does Prometheus have release channels that provide any kind of stability guarantees?

File: terraform/modules/monitoring/monitoring.tf
Comment (tgeoghegan) [inline terraform/modules/monitoring/monitoring.tf:66]: Could you include a comment explaining what these values are and why they're needed? I followed your investigation in the `gcp` channel on the ENPA slack and my takeaway was "ooh, that's a bit confusing".

File: terraform/modules/monitoring/monitoring.tf
Comment (tgeoghegan) [inline terraform/modules/monitoring/monitoring.tf:148]: This is the volume that Prometheus is going to store time series data in, right? Any thoughts on whether 5 GB or 20 GB will be sufficient? I know that's a tricky capacity planning question given that we don't even yet know what kind of data we want to store here, but maybe we should make these parameters configurable so it's easy to grow these volumes in the future?

File: terraform/modules/monitoring/monitoring.tf
Comment (tgeoghegan) [reply terraform/modules/monitoring/monitoring.tf:179]: Sounds like we should use a regional disk, then.

**Review by tgeoghegan — CHANGES_REQUESTED** [review]
I'm a dunce, and like having things explained to me in simple terms, so I'd love to see a discussion in the top level README explaining what we use Prometheus for and how it works, and how that impacts namespace and network topology. For instance, let's say we want `workflow-manager` to emit a data point to a metric every time it schedules an `intake-batch` job. How does that work? Are we going to have to run a Prometheus exporter in each pod that relays metrics to the central Prometheus? Or do I have it backwards and each pod will have to add a service that listens on 9090 so that the central Prometheus can scrape metrics from them? I suppose a lot of that will become obvious once we wire up the first metrics from a workflow-manager.

File: terraform/modules/monitoring/monitoring.tf
Comment (jsha) [reply terraform/modules/monitoring/monitoring.tf:49]: For DockerHub images, `latest` generally means "latest release." And Prometheus releases are pretty stable over time - no big API breaks. I think we're better off staying up-to-date with `latest` than pinning a version. Also, this is one of the recommended install methods: https://prometheus.io/docs/prometheus/latest/installation/

File: terraform/modules/monitoring/monitoring.tf
Comment (jsha) [reply terraform/modules/monitoring/monitoring.tf:148]: These are total shots in the dark. My plan is to deploy and measure how much we're using. As far as configurability - I think it's pretty easy to edit the hardcoded value here in the future, or turn it into a var.

One potential stumbling block: The "claim" is actually at 5Gi rather than the full 20Gi of its persistent disk. During testing I set it to be much lower, worried that the claim wasn't becoming bound because for some reason it was asking for too much. Later I tried increasing the claim, and ran into some errors that it couldn't be resized. Seems like we may need to do some extra stuff to allow resizes.

File: terraform/modules/monitoring/monitoring.tf
Comment (tgeoghegan) [reply terraform/modules/monitoring/monitoring.tf:49]: Thanks for double checking

**Comment by jsha** [comment]
Closing in favor of #230.

**Pull request closed without merging**
